Base/Python/slicer/parameterNodeWrapper.py

```diff
@@ -0,0 +1,814 @@
+        self._saveList()
+
+    def extend(self, other) -> None:
+        self._list.extend(other)
+        self._saveList()
+
+    def insert(self, index, item) -> None:
+        self._list.insert(index, item)
+        self._saveList()
+
+    def remove(self, item) -> None:
+        self._list.remove(item)
+        self._saveList()
+
+    def pop(self, i=-1):
+        val = self._list.pop(i)
+        self._saveList()
+        return val
+
+    def clear(self, ) -> None:
+        self._list.clear()
+        self._saveList()
+
+    def sort(self, *args, **kwargs) -> None:
+        self._list.sort(*args, **kwargs)
+        self._saveList()
+
+    def reverse(self) -> None:
+        self._list.reverse()
+        self._saveList()
+
+
+class ListSerializer(Serializer):
+    """
+    Serializer for lists of a type.
+    """
+    def __init__(self, elementTypeSerializer):
+        """
+        Constructs a ListSerializer. The elements will be serialized/deserialized with the
+        given elementTypeSerializer.
+        """
+        self._elementSerializer = elementTypeSerializer
+        self._lenSerializer = NumberSerializer(int)
+
+    def _lenName(self, name):
+        return f"{name}_len"
+
+    def _paramName(self, name, index):
+        return f"{name}_{index}"
+
+    def _len(self, parameterNode, name) -> int:
+        if self.isIn(parameterNode, name):
+            return self._lenSerializer.read(parameterNode, self._lenName(name))
+        else:
+            return 0
+
+    def _setLen(self, parameterNode, name, length):
+        self._lenSerializer.write(parameterNode, self._lenName(name), length)
+
+    def isIn(self, parameterNode, name: str) -> bool:
+        return self._lenSerializer.isIn(parameterNode, self._lenName(name))
+
+    def write(self, parameterNode, name: str, values) -> None:
+        paramName = lambda index: self._paramName(name, index)
+        with slicer.util.NodeModify(parameterNode):
+            oldLen = self._len(parameterNode, name)
+            newLen = len(values)
+
+            # unset any items that we no longer have indices for
+            for index in range(newLen, oldLen):
+                self._elementSerializer.remove(parameterNode, paramName(index))
+
+            self._setLen(parameterNode, name, newLen)
+            for index, value in enumerate(values):
+                self._elementSerializer.write(parameterNode, paramName(index), value)
+
+    def read(self, parameterNode, name):
+        paramName = lambda index: self._paramName(name, index)
+
+        ret = [self._elementSerializer.read(parameterNode, paramName(index))
+               for index in range(self._len(parameterNode, name))]
+        return ObservedList(parameterNode, self, name, ret)
+
+    def remove(self, parameterNode, name: str) -> None:
+        """Removes this parameter from the node if it exists."""
+        paramName = lambda index: self._paramName(name, index)
+        for index in range(self._len(parameterNode)):
```

Comment:
Static analyzer revealed that `self._len` and `self._lenSerializer.remove` are both missing the `name` argument here.

However I'm having a hard time working the call stack to see if/how to trigger this method to see if it's actually a problem.

Reply:
I will look into this. The way to trigger it should be to do a `list[list[something]]` and make the outer list smaller by setting it to smaller list

```
param.listlistx = [[1],[2]]
param.listlistx = [[3]]
```

Reply:
Ah, yes it is an issue.

Suggest to add to `test_list_of_lists`:


  pl.pop(0)
  self.assertEqual(param.p, [[4, 5, 0, 7]])
  self.assertEqual(param2.p, [[4, 5, 0, 7]])

Might also consider overriding `ObservableList.__del__` to support `del pl[0]`; this is what I tried before but it does not propagate the change to the serializer.

Reply:
Will add the test. Will look int overriding `__del__`.

Reply:
Re: __del__

I think that `__delitem__` is the right thing here, which I implemented, but the implementation was wrong (it didn't save the updated list).

Reply:
See https://github.com/Slicer/Slicer/pull/6684/commits/e3b8b690cbc93be711f1bf69fd53384b106a19eb

Reply:
See also https://github.com/Slicer/Slicer/pull/6684/commits/638e6a5a0f513caccc617fa09513e522be40236f

Reply:
Oops! Yes, `__delitem__` is what I meant.